File: Best_Time_to_Buy_and_Sell_Stock_with_Transaction_Fee.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices, int fee) {
        int n = prices.size();
        if (n < 2) {
            return 0;
        }

        int buy = -prices[0];
        int sell = 0;

        for (int i = 1; i < n; i++) {
            int temp = buy;
            buy = std::max(buy, sell - prices[i]);
            sell = std::max(sell, temp + prices[i] - fee);
        }

        return sell;        
    }
};

File: Best_Time_to_Buy_and_Sell_Stock_with_Transaction_Fee.py
class Solution:
    def maxProfit(self, prices: List[int], fee: int) -> int:
        n = len(prices)
        if n < 2:
            return 0

        buy = -prices[0]
        sell = 0

        for i in range(1, n):
            temp = buy
            buy = max(buy, sell - prices[i])
            sell = max(sell, temp + prices[i] - fee)
        
        return sell
